### PR TITLE
First draft of 4.2 update

### PR DESCRIPTION
**Pull request by marmarta**
First draft of update script from 4.1 to 4.2

QubesOS/qubes-issues#7832

### CODE REVIEW

File: qubes-dist-upgrade.sh [hunk 1]
```diff
@@ -474,7 +109,7 @@ if [[ $EUID -ne 0 ]]; then
    exit 1
 fi
 
-if ! OPTS=$(getopt -o htrlsgydu:n:f:jkp --long help,all,update,template-standalone-upgrade,release-upgrade,dist-upgrade,setup-efi-grub,assumeyes,double-metadata-size,usbvm:,netvm:,updatevm:,skip-template-upgrade,skip-standalone-upgrade,resync-appmenus-features,post-reboot,only-update:,max-concurrency:,keep-running: -n "$0" -- "$@"); then
+if ! OPTS=$(getopt -o htrlsgydu:n:f:jkp --long help,all,update,template-standalone-upgrade,release-upgrade,dist-upgrade,assumeyes,usbvm:,netvm:,updatevm:,skip-template-upgrade,skip-standalone-upgrade,resync-appmenus-features,post-reboot,only-update:,max-concurrency:,keep-running: -n "$0" -- "$@"); then
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: drop also short options that don't exist anymore

File: scripts/upgrade-template-standalone.sh [hunk 2]
```diff
@@ -18,42 +17,36 @@ fi
...
+    if [ "$releasever" -lt 11 ]; then
         exit 2
     fi
-    # Backup R4.0 repository file
+    # Backup R4.1 repository file
     cp /etc/apt/sources.list.d/qubes-r4.list /etc/apt/sources.list.d/qubes-r4.list.bak
     # We don't have $releasever into so manually replace it
-    sed -i 's/r4.0/r4.1/g' /etc/apt/sources.list.d/qubes-r4.list
+    sed -i 's/r4.1/r4.2/g' /etc/apt/sources.list.d/qubes-r4.list
+    sed -i 's/arch=amd64/arch=amd64\ signed-by=\/usr\/share\/keyrings\/apt-qubes-archive-keyring.gpg/g' /etc/apt/sources.list.d/qubes-r4.list
```
Comment (marmarek) [inline scripts/upgrade-template-standalone.sh]:
```suggestion
    sed -i 's/arch=amd64/arch=amd64\ signed-by=\/usr\/share\/keyrings\/apt-qubes-archive-keyring-4.2.gpg/g' /etc/apt/sources.list.d/qubes-r4.list
```

File: qubes-dist-upgrade.sh [hunk 3]
```diff
@@ -748,14 +296,8 @@ if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
...
+        packages="qubes-audio-daemon qubes-gui-daemon qubes-artwork-plymouth"
 
-        # Install new Qubes Grub theme before not being able to
-        # download anything else due to distro-sync
-        packages="$packages grub2-qubes-theme qubes-artwork-plymouth"
-        if is_qubes_uefi; then
-            packages="$packages grub2-efi-x64"
-        fi
         # shellcheck disable=SC2086
         qubes-dom0-update $dnf_opts --downloadonly $packages
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: Add `--force-xen-upgrade` to avoid the prompt to shutdown all VMs.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: scripts/upgrade-template-standalone.sh [hunk 4]
```diff
@@ -18,42 +17,38 @@ fi
...
     # Ensure DNF cache is cleaned
     dnf clean all
     # Run upgrade
     if ! dnf distro-sync -y --best --allowerasing; then
         exit 3
     fi
-    # Removing xen-qubes-vm pkg disables this service, even though it's still
-    # present in the system (via xen-runtime package now).
-    systemctl preset xendriverdomain
+    dnf swap--allowerasing pulseaudio pipewire-pulseaudio
```
Comment (noskb) [inline scripts/upgrade-template-standalone.sh]:
```suggestion
    dnf swap --allowerasing pulseaudio pipewire-pulseaudio
```

**Comment by marmarek** [comment]
PipelineRetry

File: qubes-dist-upgrade.sh [hunk 5]
```diff
@@ -719,108 +324,53 @@ if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
...
+      qubesctl saltutil.clear_cache
+      qubesctl saltutil.sync_all
+
+      echo "---> (STAGE 5) Adjusting default kernel"
+      default_kernel="$(qubes-prefs default-kernel)"
+      default_kernel_path="/var/lib/qubes/vm-kernels/$default_kernel"
+      default_kernel_package="$(rpm --qf '%{NAME}' -qf "$default_kernel_path")"
+      if [ "$default_kernel_package" = "kernel-qubes-vm" ]; then
+          new_kernel=$(rpm -q --qf '%{VERSION}-%{RELEASE}\n'  kernel-qubes-vm | sort -V | tail -1)
+          new_kernel="${new_kernel%.qubes}"  # TODO: does this work? check with tests
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: No, it doesn't, there may be `6.1.38-1.qubes.fc37`. Try `new_kernel="${new_kernel/.qubes/}"`

File: qubes-dist-upgrade.sh [hunk 6]
```diff
@@ -444,8 +76,14 @@ shutdown_nonessential_vms() {
     mapfile -t running_vms < <(qvm-ls --running --raw-list --fields name)
     keep_running=( dom0 "$usbvm" "$netvm" "$updatevm" "${extra_keep_running[@]}" )
     # all the updates-proxy targets
-    mapfile -t updates_proxy < <(grep '^\s*[^#].*target=' /etc/qubes-rpc/policy/qubes.UpdatesProxy | cut -d = -f 2)
-    keep_running+=( "${updates_proxy[@]}" )
+    if [ -e "/etc/qubes-rpc/policy/qubes.UpdatesProxy" ]; then
+        mapfile -t updates_proxy < <(grep '^\s*[^#].*target=' /etc/qubes-rpc/policy/qubes.UpdatesProxy | cut -d = -f 2)
+        keep_running+=( "${updates_proxy[@]}" )
+    fi
+    if [ -e "/etc/qubes-rpc/policy/qubes.UpdatesProxy" ]; then
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: This condition isn't really applicable for new policy

File: qubes-dist-upgrade.sh [hunk 7]
```diff
@@ -444,8 +76,14 @@ shutdown_nonessential_vms() {
...
+      mapfile -t updates_proxy_new < <(grep qubes.UpdatesProxy /etc/qubes/policy.d/*policy | grep '^\s*[^#].*target=' | cut -d = -f 2)
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: Technically it should be `*.policy`

File: qubes-dist-upgrade.sh [hunk 8]
```diff
@@ -537,96 +192,30 @@ max_concurrency="${max_concurrency:-4}"
...
+      root_vol_name=$(get_root_volume_name)
+      root_group_name=$(get_root_group_name)
+      if [ -z "$root_vol_name" ]; then
+        echo "---> (STAGE 1) Skipping dom0 snapshot - no LVM volume found"
+      elif lvs "$root_group_name/Qubes41UpgradeBackup" > /dev/null 2>&1 ; then
+        echo "---> (STAGE 1) Skipping dom0 snapshot - snapshot already exists. If you want to make a snapshot anyway, remove the existing one using lvremove $root_group_name/Qubes41UpgradeBackup"
+      elif [ "$assumeyes" == "1" ] || confirm "---> (STAGE 1) Do you want to make a dom0 snapshot?"; then
+        # make a dom0 snapshot
+        lvcreate -n Qubes41UpgradeBackup -s "$root_vol_name"
+        echo "--> If upgrade to 4.2 fails, you can restore your dom0 snapshot after booting from a bootable device and using sudo lvconvert --merge $root_group_name/Qubes41UpgradeBackup. Reboot after restoration."
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: It seems I was wrong, it is possible to do that from running system, and then reboot.

File: qubes-dist-upgrade.sh [hunk 9]
```diff
@@ -719,108 +359,59 @@ if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
...
+              exit 1
+          fi
+          echo "Changing default kernel from $default_kernel to $new_kernel"
+          qubes-prefs default-kernel "$new_kernel"
+      fi
+
+      root_vol_name=$(get_root_volume_name)
+      if [ "$root_vol_name" ]; then
+        root_group_name=$(get_root_group_name)
+        lvremove "$root_group_name/Qubes41UpgradeBackup"
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: if assumeyes==1, there should be `-f` too.

**Review by marmarek — CHANGES_REQUESTED** [review]

**Review by marmarek — APPROVED** [review]

**Pull request merged**